### ISSUE TITLE
deepcopy-gen should support unset GOPATH

### ISSUE THREAD

**Issue by DirectXMan12** [kubernetes-sigs/controller-tools#120]
deepcopy-gen currently requires gopath to be set, despite the fact that a default gopath exists.  We should support both that use case, and the eventual go modules usecase, which doesn't require a gopath.

cc kubernetes-sigs/kubebuilder#542

**Comment by andreykaipov** [comment]
I don't think deepcopy-gen is an issue actually.

Following https://github.com/kubernetes-sigs/controller-tools/pull/175, I was trying to migrate Kubebuilder to use Go modules too. The generation of the DeepCopy methods is fine, but the tricky part is trying to get the generator to place the file in the proper place.

Instead of specifying `./...` as our import path (see https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/scaffold/manager/apis.go#L37), we just need to specify it from the module name, e.g `github.com/abc/def/pkg/apis/...`, assuming our module is called `github.com/abc/def`. This successfully generates the deepcopy file!

The issue with the above approach is gengo places the generated file based on the import path. So, if we specify an output base dir of `.`, gengo will create the file under `./github.com/abc/def/pkg/apis/<type>/<version>/zz_generated.deepcopy.go`. This isn't ideal, since we're generating the file from inside of the project.

There's actually a gengo flag `--output-package` that looks like it's meant to override this behavior, but it's not actually used anywhere in the library, so it doesn't help us. Even if it did behave like this, we wouldn't be able to use it in the `go:generate` comment since the type and version of the API resource varies.

I think a nice approach would be to add a flag to gengo that trims a given prefix from the input paths. That way, the final `go:generate` script can look like this:

```
//go:generate go run ../../vendor/k8s.io/code-generator/cmd/deepcopy-gen/main.go -O zz_generated.deepcopy -i github.com/abc/def/pkg/apis/... -t github.com/abc/def -o .
```

**Comment by DirectXMan12** [comment]
this is on hold till we finish revamping controller-gen.  Ideally, we'll have deepcopy-gen built-in afterwards.

**Comment by andreykaipov** [comment]
Oh like built into this library so Kubebuilder projects won't have to vendor code-generator's `deepcopy-gen`?

**Comment by DirectXMan12** [comment]
yeah, exactly

**Comment by DirectXMan12** [comment]
fixed in v0.2.0-xyz
